lib/hawkular/operations/operations_api.rb

```diff
@@ -79,22 +83,34 @@ def initialize(args)
 
       super(args[:host], args[:credentials], args[:options])
 
-      url = "ws#{args[:use_secure_connection] ? 's' : ''}://#{args[:host]}/hawkular/command-gateway/ui/ws"
+      @logger = Hawkular::Logger.new
+
+      @url = "ws#{args[:use_secure_connection] ? 's' : ''}://#{args[:host]}/hawkular/command-gateway/ui/ws"
+      @credentials = args[:credentials]
+      @tenant = args[:options][:tenant]
+      @wait_time = args[:wait_time]
+    end
+
+    def base64_credentials
+      ["#{@credentials[:username]}:#{@credentials[:password]}"].pack('m').delete("\r\n")
+    end
 
-      creds = args[:credentials]
-      base64_creds = ["#{creds[:username]}:#{creds[:password]}"].pack('m').delete("\r\n")
+    def connect
+      return if @connecting || (@ws && @ws.open?)
+
+      @connecting = true
+
+      condition = new_cond
```

Comment:
It comes from `MonitorMixin`, http://ruby-doc.org/stdlib-2.4.1/libdoc/monitor/rdoc/MonitorMixin.html#method-i-new_cond , and generates a Mutex that is used later with `wait_until` to keep it locked until we finish connecting.